pepper-apis/ddp-common/src/main/java/org/broadinstitute/ddp/client/GoogleBucketClient.java

```diff
@@ -76,4 +87,73 @@ public void deleteBlob(Blob blob) {
+                .setPayloadFormat(PayloadFormat.JSON_API_V1)
+                .setEventTypes((EventType[])eventTypes.toArray())
+                .build();
+        
+        /*
+         * Due to how Google handles notifications in Cloud Storage, the method needs
+         *  to do a bit of housekeeping as well when enabling notifications.
+         * 
+         * Google Storage considers each notification to be distinct, and there may be
+         *  multiple notifications which have the same topic and event types. This implentation
+         *  assumes that enableNotifications() will handle the subscription maintenance, and clear
+         *  any non-matching notifications.
+         *  
+         */
+        var notifications = fetchAllNotifications(bucket, topic);
+        var expectedEventTypes = Optional.ofNullable(notificationInfo.getEventTypes())
+                .map(events -> new HashSet<>(events))
+                .orElse(new HashSet<>());
+
+        final var notificationsToDelete = new HashSet<String>();
+        
+        /*
+         * Any notifications which do not match the desired set of event types should be removed
+         */
+        notifications.stream()
+                .filter(notification -> {
+                    final var notificationEvents = Optional.ofNullable(notification.getEventTypes())
+                            .map(events -> new HashSet<>(events))
+                            .orElse(new HashSet<>());
+                    return !notificationEvents.equals(expectedEventTypes);
+                })
+                .map(notification -> notification.getNotificationId())
+                .forEach(notificationId -> notificationsToDelete.add(notificationId));
+
+        /*
+         * Notifications which do not match the desired payload format should also be removed
+         */
+        notifications.stream()
+                .filter(notification -> notification.getPayloadFormat() != notificationInfo.getPayloadFormat())
+                .map(notification -> notification.getNotificationId())
+                .forEach(notificationId -> notificationsToDelete.add(notificationId));
+        
+        notifications.removeIf(notification -> notificationsToDelete.contains(notification.getNotificationId()));
+
+        notificationsToDelete.forEach(notificationId -> storage.deleteNotification(bucket, notificationId));
+
+        if (notifications.isEmpty() == false) {
+            /*
+             * At least one notification exists with the necessary properties.
+             * This does _not_ preclude other notifications with different ETags,
+             * and other attributes.
+             * 
+             * None of the additional attributes are being used as of this time.
+             */
+            return true;
+        }
+
+        var createdNotification = storage.createNotification(bucket, notificationInfo);
+        return (createdNotification != null);
```

Comment:
The intent of this block is to allow for automated setup of the necessary Cloud Storage notifications. Since there is no identifier for these notifications being persisted, heuristics needs to be used to determine if the existing subscriptions (if there are any) are needed or not in an effort to prevent creating multiple notifications.

I'm on the fence about this code as, while it does ease setup, it introduces risk in managing the notifications (including potential privilege limitations).

See the description of the PR for the manual process for setting up the required notification with `gsutil`.